x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/utils/expand_dotted.test.ts

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { expandDottedObject } from './expand_dotted';
+
+describe('Expand Dotted', () => {
+  it('expands simple dotted fields to nested objects', () => {
+    const simpleDottedObj = {
+      'kibana.test.1': 'the spice must flow',
+      'kibana.test.2': 2,
+      'kibana.test.3': null,
+    };
+    expect(expandDottedObject(simpleDottedObj)).toEqual({
+      kibana: {
+        test: {
+          1: 'the spice must flow',
+          2: 2,
+          3: null,
+        },
+      },
+    });
+  });
+
+  it('expands complex dotted fields to nested objects', () => {
+    const complexDottedObj = {
+      'kibana.test.1': 'the spice must flow',
+      'kibana.test.2': ['a', 'b', 'c', 'd'],
+      'kibana.test.3': null,
+      'signal.test': {
+        key: 'val',
+      },
+      'kibana.alert.ancestors': [
+        {
+          name: 'ancestor1',
+        },
+        {
+          name: 'ancestor2',
+        },
+      ],
+      flat: 'yep',
+    };
+    expect(expandDottedObject(complexDottedObj)).toEqual({
+      kibana: {
+        alert: {
+          ancestors: [
+            {
+              name: 'ancestor1',
+            },
+            {
+              name: 'ancestor2',
+            },
+          ],
+        },
+        test: {
+          1: 'the spice must flow',
+          2: ['a', 'b', 'c', 'd'],
+          3: null,
+        },
+      },
+      signal: {
+        test: {
+          key: 'val',
+        },
+      },
+      flat: 'yep',
+    });
+  });
+});
```

Comment:
Runtime fields can also produce non-expandable objects, right?

E.g. 
```json
{
  "a.b": "value",
  "a": "runtime_value",
}
```

Reply:
Runtime fields can produce non-expandable objects, but this function is only used (for now) on alerts that are index-able into Elasticsearch. When this function is called the merge-able runtime fields will have already been merged in by the merge strategy logic used to build alerts from search results. The non-mergeable runtime fields won't be included here. (if the non-mergeable runtime fields reach this point, then they're part of the document that we're going to index into Elasticsearch. However, Elasticsearch rejects non-mergeable fields with the error below, so we'll have additional problems in that case)
![image](https://user-images.githubusercontent.com/55718608/139518112-4dbae761-8ca6-4503-9ddc-9fc4d54f0ddd.png)

Reply:
Well, the alerts are supposed to be indexable. I just noticed that a source document that contains the field `kibana.alert: "test_value"` causes an invalid alert document to be built since it conflicts with the `kibana.alert` object we use.

We really need to put some more time into standardizing the way we handle field accesses and document building, as the number of edge cases right now is huge and the multiple notations make it too hard to keep track of IMO.